vulnerable_people_form/form_pages/shared/routing.py

```diff
@@ -262,70 +272,77 @@ def dynamic_back_url(default="/"):
 
     return append_querystring_params(back_url)
 
 
 def get_redirect_for_returning_user_based_on_tier():
     original_postcode = form_answers()["support_address"]["postcode"]
-    original_postcode_tier = get_postcode_tier()
+    original_uprn = form_answers()["support_address"].get("address_uprn", None)
+    original_location_tier = get_location_tier()
+
+    latest_location_tier = None
+    if original_uprn:
+        latest_location_tier = location_eligibility.get_uprn_tier(original_uprn)
+    else:
+        latest_location_tier = location_eligibility.get_postcode_tier(original_postcode)
 
-    latest_postcode_tier_info = get_latest_postcode_tier(original_postcode, original_postcode_tier)
+    latest_location_tier_info = get_latest_location_tier(latest_location_tier, original_location_tier)
 
-    if latest_postcode_tier_info is None:
+    if latest_location_tier_info is None:
         return redirect("/not-eligible-postcode-returning-user-tier-not-found")
 
-    latest_postcode_tier = PostcodeTier(latest_postcode_tier_info["latest_postcode_tier"])
-    set_postcode_tier(latest_postcode_tier)
+    latest_location_tier = PostcodeTier(latest_location_tier_info["latest_location_tier"])
+    set_location_tier(latest_location_tier)
 
-    postcode_tier_change_status = PostcodeTierStatus(latest_postcode_tier_info["change_status"])
+    location_tier_change_status = PostcodeTierStatus(latest_location_tier_info["change_status"])
 
-    if postcode_tier_change_status == PostcodeTierStatus.NO_CHANGE:
+    if location_tier_change_status == PostcodeTierStatus.NO_CHANGE:
         return get_redirect_to_terminal_page()
-    elif postcode_tier_change_status == PostcodeTierStatus.INCREASED:
-        if latest_postcode_tier == PostcodeTier.VERY_HIGH_PLUS_SHIELDING:
+    elif location_tier_change_status == PostcodeTierStatus.INCREASED:
+        if latest_location_tier == PostcodeTier.VERY_HIGH_PLUS_SHIELDING:
             return redirect("/basic-care-needs?ca=1")
-        _raise_returning_user_redirect_error(postcode_tier_change_status, original_postcode_tier, latest_postcode_tier)
-    elif postcode_tier_change_status == PostcodeTierStatus.DECREASED:
-        if latest_postcode_tier == PostcodeTier.VERY_HIGH:
+        _raise_returning_user_redirect_error(location_tier_change_status, original_location_tier, latest_location_tier)
+    elif location_tier_change_status == PostcodeTierStatus.DECREASED:
+        if latest_location_tier == PostcodeTier.VERY_HIGH:
             return get_redirect_to_terminal_page()
-        elif not is_tier_very_high_or_above(latest_postcode_tier):
+        elif not is_tier_very_high_or_above(latest_location_tier):
             return redirect("/not-eligible-postcode-returning-user")
-        _raise_returning_user_redirect_error(postcode_tier_change_status, original_postcode_tier, latest_postcode_tier)
+        _raise_returning_user_redirect_error(location_tier_change_status, original_location_tier, latest_location_tier)
```

Comment:
I'm slightly concerned about all this nested conditional logic at a time when new levels are being introduced but that can easily argue for getting this in asap